core/internal/filetransfer/file_transfer_s3.go

```diff
@@ -0,0 +1,289 @@
+	"golang.org/x/sync/errgroup"
+)
+
+var maxS3Workers int = 500
+var s3Scheme string = "s3"
+
+// S3FileTransfer uploads or downloads files to/from s3
+type S3FileTransfer struct {
+	// client is the HTTP client for the file transfer
+	client *s3.Client
+
+	// logger is the logger for the file transfer
+	logger *observability.CoreLogger
+
+	// fileTransferStats is used to track upload/download progress
+	fileTransferStats FileTransferStats
+
+	ctx context.Context
+}
+
+// News3FileTransfer creates a new fileTransfer
+func NewS3FileTransfer(
+	client *s3.Client,
+	logger *observability.CoreLogger,
+	fileTransferStats FileTransferStats,
+) (*S3FileTransfer, error) {
+	ctx := context.TODO()
+	if client == nil {
+		var awsConfigOptions [](func(*config.LoadOptions) error)
+		if awsProfile := os.Getenv("AWS_PROFILE"); awsProfile != "" {
+			awsConfigOptions = append(
+				awsConfigOptions,
+				config.WithSharedConfigProfile(awsProfile),
+			)
+		}
+		if awsRegion := os.Getenv("AWS_REGION"); awsRegion != "" {
+			awsConfigOptions = append(awsConfigOptions, config.WithRegion(awsRegion))
+		}
+		cfg, err := config.LoadDefaultConfig(ctx, awsConfigOptions...)
+		if err != nil {
+			return nil, err
+		}
+		client = s3.NewFromConfig(cfg)
+	}
+
+	fileTransfer := &S3FileTransfer{
+		logger:            logger,
+		client:            client,
+		fileTransferStats: fileTransferStats,
+		ctx:               ctx,
+	}
+	return fileTransfer, nil
+}
+
+// Upload uploads a file to the server
+func (ft *S3FileTransfer) Upload(task *ReferenceArtifactUploadTask) error {
+	ft.logger.Debug("S3 file transfer: uploading file", "path", task.PathOrPrefix)
+
+	return nil
+}
+
+// Download downloads a file from the server
+func (ft *S3FileTransfer) Download(task *ReferenceArtifactDownloadTask) error {
+	ft.logger.Debug("s3 file transfer: downloading file", "path", task.PathOrPrefix, "ref", task.Reference)
+
+	// Parse the reference path to get the scheme, bucket, and object
+	bucketName, rootObjectName, err := parseCloudReference(task.Reference, s3Scheme)
+	if err != nil {
+		return ft.formatDownloadError("error parsing reference", err)
+	}
+
+	var getObjectInputs []*s3.GetObjectInput
+	if task.HasSingleFile() {
+		getObjInput, err := ft.getObjectInputFromTask(bucketName, rootObjectName, task)
+		if err != nil {
+			return ft.formatDownloadError("error constructing object input", err)
+		}
+		getObjectInputs = []*s3.GetObjectInput{getObjInput}
+	} else {
+		getObjectInputs, err = ft.listObjectsWithPrefix(bucketName, rootObjectName)
+		if err != nil {
+			return ft.formatDownloadError(
+				fmt.Sprintf(
+					"error getting objects in bucket %s under prefix %s",
+					bucketName, rootObjectName,
+				),
+				err,
+			)
+		}
+	}
+	return ft.downloadFiles(bucketName, rootObjectName, getObjectInputs, task.PathOrPrefix)
+}
+
+// getObjectFromTask finds the s3 object that matches the versionId and
+// digest/ETag specified by the task
+func (ft *S3FileTransfer) getObjectInputFromTask(
```

Comment:
`GetObjectInput` is a _terrible_ name for a type--not your fault, shame on AWS. I can see going either way, but it might be clearer to depart from their naming scheme and name this function something more informative about what it's actually doing